### PR TITLE
Introduce replaceRoutes() method and 2 new constructors to RestHandler.java

### PR DESCRIPTION
**Pull request by afazel**
### Description
The purpose of this PR is to add a method to `RestHandler` interface so it can add prefixes to both old and new routes for REST APIs.
 
### Issues Resolved
The PR will address the request from @vrozov in PR #1256: https://github.com/opensearch-project/security/pull/1256
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ Done] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ Done] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by vrozov — COMMENTED** [review]
nit: add unit test

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (afazel) [reply server/src/main/java/org/opensearch/rest/RestHandler.java]: Done

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (afazel) [reply server/src/main/java/org/opensearch/rest/RestHandler.java]: Done

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (afazel) [reply server/src/main/java/org/opensearch/rest/RestHandler.java]: Done

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (afazel) [reply server/src/main/java/org/opensearch/rest/RestHandler.java]: Why should we define 2 methods for ReplaceRoute?

Comment (vrozov) [reply server/src/main/java/org/opensearch/rest/RestHandler.java]: For convenience. It is more common that `method` will be the same for new and deprecated paths.

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (vrozov) [reply server/src/main/java/org/opensearch/rest/RestHandler.java]: It was a comment, not a suggestion.

**Comment by afazel** [comment]
> nit: add unit test

What is the best place to add unit tests? I am not familiar with `OpenSearch` code base that much.
Also having unit tests in different plugins, like security is not enough?

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (vrozov) [inline server/src/main/java/org/opensearch/rest/RestHandler.java:168]:
```suggestion
            this(route.getMethod, prefix + route.getPath(), deprecatedPrefix + route.getPath()
```
and introduce 2 ctor.

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (vrozov) [inline server/src/main/java/org/opensearch/rest/RestHandler.java:183]: Consider: "Construct replaced routes using routes template and prefixes for new and deprecated paths"

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (vrozov) [inline server/src/main/java/org/opensearch/rest/RestHandler.java:188]: I don't think it is necessary to clarify that number of routes will be the same (it was not a suggestion).

Comment (afazel) [reply server/src/main/java/org/opensearch/rest/RestHandler.java:188]: Got it, thanks Vlad.

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (afazel) [reply server/src/main/java/org/opensearch/rest/RestHandler.java]: new ctor added in the latest commit.

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (vrozov) [inline server/src/main/java/org/opensearch/rest/RestHandler.java:168]:
```suggestion
            this(route.getMethod(), prefix + route.getPath(), deprecatedPrefix + route.getPath());
```
use newly introduced ctor

**Review by vrozov — APPROVED** [review]
LGTM

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (afazel) [reply server/src/main/java/org/opensearch/rest/RestHandler.java]: Done

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (afazel) [reply server/src/main/java/org/opensearch/rest/RestHandler.java]: Done

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (saratvemulapalli) [inline server/src/main/java/org/opensearch/rest/RestHandler.java:167]: Minor nit: Thanks for adding these constructors, they seem subtle differences, can you add java doc comments for all the 3 of them?

**Review by saratvemulapalli — COMMENTED** [review]
Thanks for this change. 
Could you add a unit test for this? You should be able to add one here: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (dblock) [reply server/src/main/java/org/opensearch/rest/RestHandler.java:167]: Minor: add a space before `{`

**Comment by dblock** [comment]
> nit: add unit test

Not a nit for me. New code needs new tests. Thanks ;)

File: server/src/main/java/org/opensearch/rest/RestHandler.java
Comment (afazel) [reply server/src/main/java/org/opensearch/rest/RestHandler.java]: Sure, done

**Comment by vrozov** [comment]
> > nit: add unit test
> 
> Not a nit for me. New code needs new tests. Thanks ;)

The original class does not have unit tests, so nit for me. Anyway,  it is up to project maintainers to decide whether unit tests is required before PR is merged.

**Comment by saratvemulapalli** [comment]
start gradle check

**Comment by saratvemulapalli** [comment]
@afazel can you take a look the failure?
you should be able to run `./gradlew check` on your machine.

**Comment by vrozov** [comment]
> @afazel can you take a look the failure?
> you should be able to run `./gradlew check` on your machine.

@nknize can you take a look. Seems to be related to #928 after 1.0.0 release.

**Comment by dblock** [comment]
start gradle check

**Comment by afazel** [comment]
> @afazel can you take a look the failure?
> you should be able to run `./gradlew check` on your machine.

I think the issue is related to the new unit test I added, I am debugging it now

**Comment by saratvemulapalli** [comment]
start gradle check

File: server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java
Comment (vrozov) [inline server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java:393]: If there is the test case for different paths there must be a test case for different methods as well.

File: server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java
Comment (vrozov) [inline server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java:397]: Why the assert is limited to path only? Should it check method and deprecated path as well?

**Comment by saratvemulapalli** [comment]
Not sure why the gradle check fails. 
It passes on local. 

Will open an issue to track the bwc flaky test. 
I'll kick off gradle check again.

**Comment by saratvemulapalli** [comment]
start gradle check

File: server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java
Comment (afazel) [reply server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java]: Done

File: server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java
Comment (afazel) [reply server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java:397]: Done

File: server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java
Comment (vrozov) [inline server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java:347]: nit: either use `new Route()` (preferable) or `List<RestHandler.Route>` (mixing both is a bad style).

File: server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java
Comment (vrozov) [inline server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java:353]: code style: separate arguments with ", " (add space after comma)

File: server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java
Comment (vrozov) [inline server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java:352]: order of arguments to `assertEquals` needs to be reversed. (`assertEquals(Object expected, Object actual)`)

Comment (afazel) [reply server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java]: fixed

File: server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java
Comment (afazel) [reply server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java]: fixed

File: server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java
Comment (afazel) [reply server/src/test/java/org/opensearch/rest/BaseRestHandlerTests.java]: Done

**Review by vrozov — APPROVED** [review]

**Review by saratvemulapalli — APPROVED** [review]
The changes look good to me!

Thanks @afazel for all the work.

**Comment by saratvemulapalli** [comment]
start gradle check

**Comment by cliu123** [comment]
@dblock @saratvemulapalli Are we going to backport this to 1.0 or 1.x(for 1.1.0) branches?

**Comment by dblock** [comment]
Since this is not a security fix it definitely doesn't go to 1.0, but I don't see why we wouldn't backport to 1.1. Feel free to make that PR @afazel if you need it.

**Pull request merged**
